File: go/003.go
package main

import (
	"fmt"
	"time"
)

var t time.Time

var nextPrime = func() (func() int64) {
	cur    := int64(1)
	var primes []int64;

	isPrime := func(n int64) bool {
		for _, prime := range primes {
			if n % prime == 0 {
				return false
			}
		}
		return true
	}

	return func() int64 {
		for cur++; !isPrime(cur); cur++ {}
		primes = append(primes, cur)
		if (len(primes) % 1000 == 0) {
			fmt.Print(time.Since(t))
			fmt.Print("\tprimes now has ", len(primes))
			fmt.Println("\tlargest prime: ", primes[len(primes) - 1])
		}
		return cur
	}
}()

func maxPrimeFactor(n int64) int64 {
	for {
		i := nextPrime()
		for n % i == 0 {
			if n == i { return i }
			n /= i
		}
	}

	panic("Unreachable"); // Please the compiler
}

func main() {
	t = time.Now()
	//println(maxPrimeFactor(13195))
	println(maxPrimeFactor(600851475143))
}

File: r/003.r
#!/usr/bin/Rscript

nextPrime <- (function() {
    cur <- 1
    primes <- NULL

    isNextPrime <- function(n) {
        for (prime in primes) {
            if (n %% prime == 0) {
                return(FALSE)
            }
        }
        TRUE
    }

    function() {
        cur <<- cur + 1
        while (!isNextPrime(cur)) {
            cur <<- cur + 1
        }

        primes <<- append(primes, cur)

        cur
    }
})()

maxPrimeFactor <- function(n) {
    factors <- NULL

    while(TRUE) {
        i <- nextPrime()
        while (n %% i == 0) {
            factors <- append(factors, i)
            cat(" +++ Factors:", factors, "\n")
            if (n == i) { return(i) }
            n <- n / i
            cat(" * New n:", n, "\n")
        }
    }
}

cat(maxPrimeFactor(600851475143), "\n")
